### PR TITLE
Sprint8 task1 rewards delivery method

### PR DESCRIPTION
**Pull request by iAcay**
Hey! The code which is below was implemented in order to meet the acceptance criterias of the first task in sprint no. 8 :)

AC in this task:
-- A reward should have assigned a delivery method: `online`, or `post-delivery`
-- When an employee chooses online - everything should stay as it's now
-- When an employee chooses post-delivery he should be able to enter the delivery address - street, postcode, and city
-- There is a presence validation set on the delivery address data
-- An employee should see an error message once he skips one of the validated fields
-- An admin can set the delivery method for each of the rewards
-- Rewards import should be able to process the delivery method
-- Once an employee sets his delivery address that address should be used for the next `post-delivery` order

https://user-images.githubusercontent.com/96024046/185563249-a683df19-95b2-49bb-b368-0a83cc25d329.mov

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -37,4 +32,18 @@ def create
...
+  def address_params
+    params.require(:address).permit(:street, :postcode, :city) if params[:address]
+  end
+
+  def order_form_params
+    order_params.merge(address_params)
+  end
+
+  def which_order_form
+    return OrderPostForm if @reward.delivery_method == 'post_delivery'
```
Comment (ciejj) [inline app/controllers/orders_controller.rb]: @reward.delivery_method == 'post_delivery' ? OrderPostForm : OrderOnlineForm

you can also use delivery_method enum here

File: app/models/employee.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@ class Employee < ApplicationRecord
   has_many :received_kudos, class_name: 'Kudo', foreign_key: 'receiver', inverse_of: :receiver, dependent: :destroy
   has_many :orders, dependent: :destroy
   has_many :rewards, through: :orders
+  belongs_to :address, optional: true
```
Comment (ciejj) [inline app/models/employee.rb]: I would make the association another way - what if later user can have multiple addresses?

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -17,17 +17,12 @@ def create
...
-      @order.reward_snapshot = @reward
-      begin
-        ActiveRecord::Base.transaction do
-          @order.save!
-          current_employee.decrement(:earned_points, @reward.price).save!
-        end
-      rescue StandardError
-        redirect_to rewards_path, notice: 'Unfortunately something went wrong :('
-      else
+      order_form = which_order_form.new(order_form_params)
```
Comment (ciejj) [inline app/controllers/orders_controller.rb]: this action is begin to be big :) consider moving it into a service object :)

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -37,4 +32,18 @@ def create
...
+
+  def address_params
+    params.require(:address).permit(:street, :postcode, :city) if params[:address]
+  end
+
+  def order_form_params
+    order_params.merge(address_params)
+  end
+
+  def which_order_form
```
Comment (ciejj) [inline app/controllers/orders_controller.rb]: this is an interesting approach, but I would consider using one form, with some of the fields hidden if it is the delivery method, and validation of that inside the create action (service object)

**Review by ciejj — APPROVED** [review]
Interesting approach :) well done!

[edit] I have left some suggestions-  ping me if you would consider refactoring it :) I will help with service object creation :)

File: app/controllers/orders_controller.rb (hunk 1 above)
Comment (iAcay) [reply app/controllers/orders_controller.rb]: Thanks! Done 👍  My bad :D

File: app/models/employee.rb (hunk 2 above)
Comment (iAcay) [reply app/models/employee.rb]: Thanks :) I have changed direction of the association 👍

File: app/controllers/orders_controller.rb (hunk 4 above)
Comment (iAcay) [reply app/controllers/orders_controller.rb]: Thanks :) I changed it. I have implemented one form object which is responsible for both kind of delivery 👍

**Pull request merged**